lib/Sema/MiscDiagnostics.cpp

```diff
@@ -5889,3 +5889,210 @@ bool swift::diagnoseUnhandledThrowsInAsyncContext(DeclContext *dc,
+          PointerUnion<EnumElementDecl *, VarDecl *> topFieldToError,
+          DeclBaseName parentName, DescriptiveDeclKind fieldKind,
+          DeclBaseName fieldName) {
+        assert(!topFieldToError.isNull());
+        if (auto *eltDecl = topFieldToError.dyn_cast<EnumElementDecl *>()) {
+          eltDecl->diagnose(diag::moveonly_copyable_type_with_moveonly_subtype,
+                            copyableNominalType->getDescriptiveKind(),
+                            copyableNominalType->getBaseName(), fieldKind,
+                            parentName.userFacingName(),
+                            fieldName.userFacingName());
+          return;
+        }
+
+        auto *varDecl = topFieldToError.get<VarDecl *>();
+        varDecl->diagnose(diag::moveonly_copyable_type_with_moveonly_subtype,
+                          copyableNominalType->getDescriptiveKind(),
+                          copyableNominalType->getBaseName(), fieldKind,
+                          parentName.userFacingName(),
+                          fieldName.userFacingName());
+      };
+
+  SmallVector<State, 32> decls;
+  decls.emplace_back(copyableNominalType, None, nullptr);
+  while (!decls.empty()) {
+    auto state = decls.pop_back_val();
+    auto *decl = state.parentTypeDecl;
+    auto optSubMap = state.subMap;
+    auto initialTopFieldToError = state.topFieldToError;
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "Loop top! Visiting: " << decl->getName() << '\n');
+
+    // First see if the decl we are checking is an enum.
+    if (auto *enumDecl = dyn_cast<EnumDecl>(decl)) {
+      // If we have an enum but we don't have any elements, just continue, we
+      // have nothing to check.
+      if (enumDecl->getAllElements().empty())
+        continue;
+
+      // Otherwise for each element...
+      for (auto *enumEltDecl : enumDecl->getAllElements()) {
+        // If the element doesn't have any associated values, we have nothing to
+        // check, so continue.
+        if (!enumEltDecl->hasAssociatedValues())
+          continue;
+
+        LLVM_DEBUG(llvm::dbgs() << "Visiting enum elt decl: "
+                                << enumEltDecl->getName() << '\n');
+
+        // Otherwise, we have a case and need to check the types of the
+        // parameters of the case payload.
+        for (auto payloadParam : *enumEltDecl->getParameterList()) {
+          LLVM_DEBUG(llvm::dbgs() << "Visiting payload param: "
+                                  << payloadParam->getName() << '\n');
+          auto topFieldToError = initialTopFieldToError;
+
+          // First grab the type of the param type and if our parent type was
+          // generic with a substitution map, apply that substitution map. This
+          // ensures that we error properly on things like
+          // Optional<MoveOnlyType>. If we didn't, we would treat the former
+          // like Optional<T>. Since T is considered copyable, we would not
+          // error which would be incorrect.
```

Comment:
`Optional<MoveOnlyType>` should already be an error to form, because move-only types cannot be used as generic substitutions, so we don't need to do additional work to diagnose it specifically. 

If you want to plan for the future of generic types that may be conditionally copyable, though, I think a better way to do that is to put a method `isMoveOnly()` on `TypeBase`, and do `payloadParamType->isMoveOnly()`. The implementation of `TypeBase::isMoveOnly()` can for now just check the NominalTypeDecl for whether it `isMoveOnly`, and when we add generics support in the future, we can enhance the check once there.

Reply:
Removed the recursion as per our conversation